File: Extend/Constraints/CPP/Range/RangeExample.cpp
/**********************************************************************/
/*           Copyright 2005, Analytical Graphics, Inc.                */
/**********************************************************************/
#include "stdafx.h"
#include "RangeExample.h"
#include "AGI.Access.Constraint.Plugin.Examples.CPP_i.c"

using namespace std;
using namespace AccessConstraint;

CRangeExample::CRangeExample()
{
}

CRangeExample::~CRangeExample()
{
}

HRESULT CRangeExample::FinalConstruct()
{
	m_pStkPluginSite = NULL;
	m_pStkRootObject = NULL;
	m_DisplayName	 = std::string("CPP.RangeExample");

	return S_OK;
}

void CRangeExample::FinalRelease()
{
}


void CRangeExample::Message(AgEUtLogMsgType severity, std::string& msg)
{
	if(m_pStkPluginSite)
	{
		USES_CONVERSION;

		CComBSTR bstrMsg = A2W(msg.c_str());

		m_pStkPluginSite->Message( severity, bstrMsg );
	}
}

STDMETHODIMP CRangeExample::get_DisplayName(BSTR *pDisplayName)
{
	USES_CONVERSION;

	CComBSTR bstrName = A2W(m_DisplayName.c_str());

	return bstrName.CopyTo(pDisplayName);
}

STDMETHODIMP CRangeExample::Register(IAgAccessConstraintPluginResultRegister* Result )
{
	USES_CONVERSION;

	CComBSTR msg(A2W("Distance"));

	Result->put_BaseObjectType(eAircraft);
	Result->put_BaseDependency(eDependencyRelativePosVel);
	Result->put_Dimension(msg);	
	Result->put_MinValue(0.0);

	Result->put_TargetDependency (eDependencyRelativePosVel);
	Result->AddTarget(eFacility);
	Result->AddTarget(eGroundVehicle);
	Result->AddTarget(eSatellite);		
	Result->Register();

	std::string msgStr = m_DisplayName + std::string(": Register(Aircraft to Facility/GroundVehicle/Satellite)");
	msg = A2W(msgStr.c_str());

	Result->Message(eUtLogMsgInfo, msg);

	Result->put_BaseObjectType(eFacility);
	Result->ClearTargets();
	Result->AddTarget(eAircraft);
	Result->AddTarget(eSatellite);	
	Result->Register();

	msgStr = m_DisplayName + std::string(": Register(Facility to Aircraft/Satellite)");
	msg = A2W(msgStr.c_str());

	Result->Message(eUtLogMsgInfo, msg);

	Result->put_BaseObjectType(eGroundVehicle);
	Result->Register();

	msgStr = m_DisplayName + std::string(": Register(GroundVehicle to Aircraft/Satellite)");
	msg = A2W(msgStr.c_str());

	Result->Message(eUtLogMsgInfo, msg);

	Result->put_BaseObjectType(eSatellite);
	Result->ClearTargets();
	Result->AddTarget(eAircraft);
	Result->AddTarget(eFacility);
	Result->AddTarget(eGroundVehicle);	
	Result->Register();

	msgStr = m_DisplayName + std::string(": Register(Satellite to Aircraft/Facility/GroundVehicle)");
	msg = A2W(msgStr.c_str());

	Result->Message(eUtLogMsgInfo, msg);

	return S_OK;
}

STDMETHODIMP CRangeExample::Init(IAgUtPluginSite* pSite, VARIANT_BOOL* pRet)
{
	pSite->QueryInterface(&m_pStkPluginSite);
	
	Message( eUtLogMsgInfo, m_DisplayName+": Init()" );

	// Demonstrate getting ObjectModel handle

	if(m_pStkPluginSite)
	{
		//----------------------------------------------------
		// Get a pointer to the STK Object Model root object
		//----------------------------------------------------
		CComPtr<IDispatch> pDisp;
		if(SUCCEEDED(m_pStkPluginSite->get_StkRootObject(&pDisp)))
		{
			pDisp->QueryInterface(&m_pStkRootObject);
		}
	}

	*pRet = VARIANT_TRUE;

	return S_OK;
}

STDMETHODIMP CRangeExample::PreCompute(IAgAccessConstraintPluginResultPreCompute* Result,
									   VARIANT_BOOL* pRet)
{
	Message( eUtLogMsgInfo, m_DisplayName+": PreCompute()" );

	// Demonstrate using ObjectModel handle

	if(m_pStkRootObject)
	{
		CComPtr<STKObjects::IAgStkObject> pScen;

		if(SUCCEEDED(m_pStkRootObject->get_CurrentScenario(&pScen)))
		{
			CComBSTR scenName;
	
			if(SUCCEEDED(pScen->get_InstanceName(&scenName)))
			{
				USES_CONVERSION;

				std::string currentScenario = W2A(scenName);

				std::string msg = std::string("Current Scenario is ");
				msg += currentScenario;

				Message(eUtLogMsgInfo, msg);
			}
		}
	}

	*pRet = VARIANT_TRUE;

	return S_OK;
}

STDMETHODIMP CRangeExample::Evaluate( 
	IAgAccessConstraintPluginResultEval* Result, 
	IAgAccessConstraintPluginObjectData* fromObject, 
	IAgAccessConstraintPluginObjectData* toObject,
	VARIANT_BOOL* pRet)
{
	if(Result)
	{
		double range;
		if(SUCCEEDED(Result->get_LightPathRange(&range)))
		{
			if(SUCCEEDED(Result->put_Value(range)))
			{
				*pRet = VARIANT_TRUE;

				return S_OK;
			}
		}
	}

	*pRet = VARIANT_FALSE;

	return E_FAIL;
}

STDMETHODIMP CRangeExample::PostCompute(IAgAccessConstraintPluginResultPostCompute* Result,
									   VARIANT_BOOL* pRet)
{
	Message( eUtLogMsgInfo, m_DisplayName+": PostCompute()" );

	*pRet = VARIANT_TRUE;

	return S_OK;
}

STDMETHODIMP CRangeExample::Free()
{
	Message( eUtLogMsgInfo, m_DisplayName+": Free()" );

	m_pStkPluginSite = NULL;
	m_pStkRootObject = NULL;

	return S_OK;
}

/**********************************************************************/
/*           Copyright 2007, Analytical Graphics, Inc.                */
/**********************************************************************/

File: Extend/Constraints/CSharp/Range/RangeExample.cs
//=====================================================//
//  Copyright 2006-2007, Analytical Graphics, Inc.     //
//=====================================================//
using Microsoft.Win32;
using System;
using System.Collections;
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text;

using AGI.Attr;
using AGI.Plugin;
using AGI.Access.Constraint.Plugin;
using AGI.STK.Plugin;
using AGI.STKObjects;
using AGI.STKUtil;

namespace AGI.Access.Constraint.Plugin.Examples.CSharp.RangeExample
{
	/// <summary>
	/// RangeExample uses the (LightPathRange) range between objects
	/// as its metric. Of course, STK already provides Range as an Access
	/// Constraint. This example demonstrates registration rather than
	/// computational complexity.
	/// 
	/// This constraint is registered for:
	///		Aircraft to Facility/GroundVehicle/Satellite
	///		Facility to Aircraft/Satellite
	///		GroundVehicle to Aircraft/Satellite
	///		Satellite to Aircraft/Facility/GroundVehicle
	///		
	/// </summary>
	// NOTE: Generate your own Guid using Microsoft's GuidGen.exe
	[Guid("2D73A444-CA85-4b63-88F1-3187E6253DC8")]
	// NOTE: Create your own ProgId to match your plugin's namespace and name
	[ProgId("AGI.Access.Constraint.Plugin.Examples.CSharp.RangeExample")]
	// NOTE: Specify the ClassInterfaceType.None enumeration, so the custom COM Interface 
	// you created is used instead of an autogenerated COM Interface.
	[ClassInterface(ClassInterfaceType.None)]
	public class RangeExample : IAgAccessConstraintPlugin
	{
		#region Data Members
		private string				m_DisplayName = "CSharp.RangeExample";
		private AgStkPluginSite		m_Site = null;
		private AgStkObjectRoot		m_StkRootObject = null;
		#endregion
		
		public RangeExample()
		{
			Debug.WriteLine(  m_DisplayName+".RangeExample()", "Entered:" );
			Debug.WriteLine(  m_DisplayName+".RangeExample()", "Exited:" );
		}

		private void Message(AGI.Plugin.AgEUtLogMsgType severity, string msg)
		{
			if(m_Site != null)
			{
				m_Site.Message( severity, msg );
			}
		}
		
		#region IAgAccessConstraintPlugin implementation
		public string DisplayName
		{
			get
			{
				return m_DisplayName;
			}
		}

		public void Register( AgAccessConstraintPluginResultRegister Result )
		{
			Result.BaseObjectType = AgEAccessConstraintObjectType.eAircraft;
			Result.BaseDependency = (int)AgEAccessConstraintDependencyFlags.eDependencyRelativePosVel;
			Result.Dimension = "Distance";	
			Result.MinValue = 0.0;
		
			Result.TargetDependency = (int)AgEAccessConstraintDependencyFlags.eDependencyRelativePosVel;
			Result.AddTarget(AgEAccessConstraintObjectType.eFacility);
			Result.AddTarget(AgEAccessConstraintObjectType.eGroundVehicle);
			Result.AddTarget(AgEAccessConstraintObjectType.eSatellite);		
			Result.Register();

			Result.Message(AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, 
				m_DisplayName+": Register(Aircraft to Facility/GroundVehicle/Satellite)");
		
			Result.BaseObjectType = AgEAccessConstraintObjectType.eFacility;
			Result.ClearTargets();
			Result.AddTarget(AgEAccessConstraintObjectType.eAircraft);
			Result.AddTarget(AgEAccessConstraintObjectType.eSatellite);	
			Result.Register();

			Result.Message(AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, 
				m_DisplayName+": Register(Facility to Aircraft/Satellite)");

			Result.BaseObjectType = AgEAccessConstraintObjectType.eGroundVehicle;
			Result.Register();

			Result.Message(AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, 
				m_DisplayName+": Register(GroundVehicle to Aircraft/Satellite)");

			Result.BaseObjectType = AgEAccessConstraintObjectType.eSatellite;
			Result.ClearTargets();
			Result.AddTarget(AgEAccessConstraintObjectType.eAircraft);
			Result.AddTarget(AgEAccessConstraintObjectType.eFacility);
			Result.AddTarget(AgEAccessConstraintObjectType.eGroundVehicle);	
			Result.Register();

			Result.Message(AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, 
				m_DisplayName+": Register(Satellite to Aircraft/Facility/GroundVehicle)");
		}

		public bool Init( IAgUtPluginSite site )
		{
			Debug.WriteLine(  m_DisplayName+".Init()", "Entered:" );

			m_Site = (AgStkPluginSite)site;
			
			Message( AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, m_DisplayName+": Init()" );

			// Demonstrate getting ObjectModel handle

			if(m_Site != null)
			{
				//----------------------------------------------------
				// Get a pointer to the STK Object Model root object
				//----------------------------------------------------
				
				m_StkRootObject = (AgStkObjectRoot)m_Site.StkRootObject;
			}

			Debug.WriteLine(  m_DisplayName+".Init()", "Exited:" );

			return true;
		}

		public bool PreCompute( AgAccessConstraintPluginResultPreCompute Result )
		{
			Debug.WriteLine(  m_DisplayName+".PreCompute()", "Entered:" );

			Message( AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, m_DisplayName+": PreCompute()" );

			// Demonstrate using ObjectModel handle

			if(m_StkRootObject != null)
			{
				IAgStkObject scenObj = m_StkRootObject.CurrentScenario;

				if(scenObj != null)
				{
					string currentScenario = scenObj.InstanceName;
					
					Message(AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, 
						"Current Scenario is "+ currentScenario);
				}
			}

			Debug.WriteLine(  m_DisplayName+".PreCompute()", "Exited:" );

			return true;
		}

		public bool Evaluate( 
			AgAccessConstraintPluginResultEval Result, 
			AgAccessConstraintPluginObjectData fromObject, 
			AgAccessConstraintPluginObjectData toObject )
		{
			if(Result != null)
			{
				Result.Value = Result.LightPathRange;
			}

			return true;
		}
		
		public bool PostCompute(AgAccessConstraintPluginResultPostCompute Result)
		{
			Debug.WriteLine(  m_DisplayName+".PostCompute()", "Entered:" );
		
			Message( AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, m_DisplayName+": PostCompute()" );

			Debug.WriteLine(  m_DisplayName+".PostCompute()", "Exited:" );
			
			return true;
		}

		public void Free()
		{
			Debug.WriteLine(  m_DisplayName+".Free()", "Entered:" );

			Message( AGI.Plugin.AgEUtLogMsgType.eUtLogMsgInfo, m_DisplayName+": Free()" );

			Debug.WriteLine(  m_DisplayName+".Free()", "Exited:" );

			m_Site = null;
			m_StkRootObject = null;
		}

		#endregion

        #region Registration functions
        /// <summary>
        /// Called when the assembly is registered for use from COM.
        /// </summary>
        /// <param name="t">The type being exposed to COM.</param>
        [ComRegisterFunction]
        [ComVisible(false)]
        public static void RegisterFunction(Type t)
        {
            RemoveOtherVersions(t);
        }

        /// <summary>
        /// Called when the assembly is unregistered for use from COM.
        /// </summary>
        /// <param name="t">The type exposed to COM.</param>
        [ComUnregisterFunctionAttribute]
        [ComVisible(false)]
        public static void UnregisterFunction(Type t)
        {
            // Do nothing.
        }

        /// <summary>
        /// Called when the assembly is registered for use from COM.
        /// Eliminates the other versions present in the registry for
        /// this type.
        /// </summary>
        /// <param name="t">The type being exposed to COM.</param>
        public static void RemoveOtherVersions(Type t)
        {
            try
            {
                using (RegistryKey clsidKey = Registry.ClassesRoot.OpenSubKey("CLSID"))
                {
                    StringBuilder guidString = new StringBuilder("{");
                    guidString.Append(t.GUID.ToString());
                    guidString.Append("}");
                    using (RegistryKey guidKey = clsidKey.OpenSubKey(guidString.ToString()))
                    {
                        if (guidKey != null)
                        {
                            using (RegistryKey inproc32Key = guidKey.OpenSubKey("InprocServer32", true))
                            {
                                if (inproc32Key != null)
                                {
                                    string currentVersion = t.Assembly.GetName().Version.ToString();
                                    string[] subKeyNames = inproc32Key.GetSubKeyNames();
                                    if (subKeyNames.Length > 1)
                                    {
                                        foreach (string subKeyName in subKeyNames)
                                        {
                                            if (subKeyName != currentVersion)
                                            {
                                                inproc32Key.DeleteSubKey(subKeyName);
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch
            {
                // Ignore all exceptions...
            }
        }
        #endregion
	}
}
//=====================================================//
//  Copyright 2006-2007, Analytical Graphics, Inc.     //
//=====================================================//